diff --git a/index.js b/index.js
index 42b996e60ee779a2b2ec1840a428b240f8a88c3c..c1475263fa43eb38c57b33f1e81034ee44585532 100644
--- a/index.js
+++ b/index.js
@@ -11,7 +11,8 @@ import { IFileService } from '@codingame/monaco-vscode-api/vscode/vs/platform/fi
 import { FileType, FilePermission, createFileSystemProviderError, FileSystemProviderErrorCode, FileSystemProviderCapabilities, FileChangeType, FileSystemProviderError, hasFileReadStreamCapability } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files';
 export { FileChangeType, FilePermission, FileSystemProviderCapabilities, FileSystemProviderError, FileSystemProviderErrorCode, FileType } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files';
 import { Disposable, DisposableStore, toDisposable } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
-import { extUri, basename, dirname } from '@codingame/monaco-vscode-api/vscode/vs/base/common/resources';
+import { extUri, basename } from '@codingame/monaco-vscode-api/vscode/vs/base/common/resources';
+import { posix } from '@codingame/monaco-vscode-api/vscode/vs/base/common/path';
 import { Emitter, Event } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';
 import { HTMLFileSystemProvider } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/browser/htmlFileSystemProvider';
 export { HTMLFileSystemProvider } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/browser/htmlFileSystemProvider';
@@ -417,7 +418,8 @@ class RegisteredFileSystemProvider extends Disposable {
             throw createFileSystemProviderError('file exists already', FileSystemProviderErrorCode.FileExists);
         }
         const basename$1 = basename(resource);
-        const dirname$1 = dirname(resource);
+        // This in-memory tree traverses URI segments, including drive and share roots, from '/'.
+        const dirname$1 = resource.with({ path: posix.dirname(resource.path) });
         const parent = this._lookupAsDirectory(dirname$1, false);
         const directory = new RegisteredDirectory();
         const disposable = new DisposableStore();
